sql/analyze_test.go

```diff
@@ -0,0 +1,553 @@
+		}
+	}
+}
+
+func TestSplitAndExpr(t *testing.T) {
+	defer leaktest.AfterTest(t)
+
+	testData := []struct {
+		expr     string
+		expected string
+	}{
+		{`a`, `a`},
+		{`a AND b`, `a, b`},
+		{`a OR b`, `a OR b`},
+		{`(a AND b) AND (c AND (d AND e))`, `a, b, c, d, e`},
+	}
+	for _, d := range testData {
+		expr, _ := parseAndNormalizeExpr(t, d.expr)
+		exprs := splitAndExpr(expr, nil)
+		if s := exprs.String(); d.expected != s {
+			t.Errorf("%s: expected %s, but found %s", d.expr, d.expected, s)
+		}
+	}
+}
+
+func TestSimplifyExpr(t *testing.T) {
+	defer leaktest.AfterTest(t)
+
+	testData := []struct {
+		expr     string
+		expected string
+	}{
+		{`a`, `a`},
+		{`a AND b`, `a AND b`},
+		{`a OR b`, `a OR b`},
+		{`(a AND b) AND (c AND d)`, `a AND b AND c AND d`},
+		{`(a OR b) OR (c OR d)`, `a OR b OR c OR d`},
+		{`a < lower('FOO')`, `a < 'foo'`},
+		{`a < 1 AND a < 2 AND a < 3 AND a < 4 AND a < 5`, `a < 1`},
+		{`a < 1 OR a < 2 OR a < 3 OR a < 4 OR a < 5`, `a < 5`},
+		{`(a < 1 OR a > 1) AND a >= 1`, `a > 1`},
+		{`a < 1 AND (a > 2 AND a < 1)`, `false`},
+		{`a < 1 OR (a > 1 OR a < 2)`, `true`},
+		{`a < 1 AND abs(a) > 0`, `a < 1`},
+		{`a < 1 OR abs(a) > 0`, `true`},
+
+		{`a IN (1, 1)`, `a IN (1)`},
+		{`a IN (2, 3, 1)`, `a IN (1, 2, 3)`},
+
+		{`a LIKE '%foo'`, `true`},
+		{`a LIKE 'foo'`, `a = 'foo'`},
+		{`a LIKE 'foo%'`, `a >= 'foo' AND a < 'fop'`},
+		{`a LIKE 'foo_'`, `a >= 'foo' AND a < 'fop'`},
+		{`a LIKE 'bar_foo%'`, `a >= 'bar' AND a < 'bas'`},
+		{`a SIMILAR TO '.*'`, `true`},
+		{`a SIMILAR TO 'foo'`, `a = 'foo'`},
+		{`a SIMILAR TO 'foo.*'`, `a >= 'foo' AND a < 'fop'`},
+		{`a SIMILAR TO '(foo|foobar).*'`, `a >= 'foo' AND a < 'fop'`},
+	}
+	for _, d := range testData {
+		expr, _ := parseAndNormalizeExpr(t, d.expr)
+		expr = simplifyExpr(expr)
+		if s := expr.String(); d.expected != s {
+			t.Errorf("%s: expected %s, but found %s", d.expr, d.expected, s)
+		}
+	}
+}
+
+func TestSimplifyNotExpr(t *testing.T) {
+	defer leaktest.AfterTest(t)
+
+	testData := []struct {
+		expr     string
+		expected string
+		check    bool
+	}{
+		{`NOT a = 1`, `a != 1`, true},
+		{`NOT a != 1`, `a = 1`, true},
+		{`NOT a > 1`, `a <= 1`, true},
+		{`NOT a >= 1`, `a < 1`, true},
+		{`NOT a < 1`, `a >= 1`, true},
+		{`NOT a <= 1`, `a > 1`, true},
+		{`NOT a IN (1, 2)`, `a NOT IN (1, 2)`, true},
+		{`NOT a NOT IN (1, 2)`, `a IN (1, 2)`, true},
+		{`NOT a LIKE 'foo'`, `true`, false},
+		{`NOT a NOT LIKE 'foo'`, `a = 'foo'`, false},
+		{`NOT a SIMILAR TO 'foo'`, `true`, false},
+		{`NOT a NOT SIMILAR TO 'foo'`, `a = 'foo'`, false},
+		{`NOT (a != 1 AND b != 1)`, `a = 1 OR b = 1`, false},
+		{`NOT (a != 1 OR a < 1)`, `a = 1`, false},
+	}
+	for _, d := range testData {
+		expr1, qvals := parseAndNormalizeExpr(t, d.expr)
+		expr2 := simplifyExpr(expr1)
+		if s := expr2.String(); d.expected != s {
+			t.Errorf("%s: expected %s, but found %s", d.expr, d.expected, s)
+		}
+		if d.check && !checkEquivExpr(t, expr1, expr2, qvals) {
+			continue
```

Comment:
`checkEquivExpr` returns false whenever it calls `t.Errorf`. I want `checkEquivExpr` to call `t.Errorf` because it can give a more explicit error message.